KAFKA-13488: Producer fails to recover if topic gets deleted midway

prat0318 opened the pull request:
Allow LeaderEpoch to be re-assigned to the new value from the Metadata Response if oldTopicId is not present in the cache. This is needed because oldTopicId is removed from the cache if the topic gets deleted but the LeaderEpoch is not removed. Hence, metadata for the newly recreated topic won't be accepted unless we allow oldTopicId to be null.

This is a fix on top of earlier made #10952 and #11004 PRs but still don't solve the bug mentioned in KAFKA-13488. This is now fixed in this PR.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

prat0318 committed:
KAFKA-13488: Producer fails to recover if topic gets deleted midway

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -394,7 +394,8 @@ else if (metadata.error() == Errors.TOPIC_AUTHORIZATION_FAILED)
             Integer currentEpoch = lastSeenLeaderEpochs.get(tp);
-            if (topicId != null && oldTopicId != null && !topicId.equals(oldTopicId)) {
+            // oldTopicId can be null (when metadata is fetched during topic recreation), update the metadata in that case as well.
+            if (topicId != null && !topicId.equals(oldTopicId)) {
```
I think this approach is reasonable. The root of the problem is the loose binding between the leader epoch and the topicId. I'm still a bit tempted to take this a little further and only permit the epoch check when the topicIds are matching. This is a little difficult to do at the moment because the epoch may be learned in contexts where we have not yet exposed the topicId. Hopefully this can be tightened up in https://issues.apache.org/jira/browse/KAFKA-13447. For now, this workaround seems ok.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -394,7 +394,8 @@ else if (metadata.error() == Errors.TOPIC_AUTHORIZATION_FAILED)
             int newEpoch = partitionMetadata.leaderEpoch.get();
             Integer currentEpoch = lastSeenLeaderEpochs.get(tp);
-            if (topicId != null && oldTopicId != null && !topicId.equals(oldTopicId)) {
+            // oldTopicId can be null (when metadata is fetched during topic recreation), update the metadata in that case as well.
```
How about this?
```java
// Between the time that a topic is deleted and re-created, the client may lose
// track of the corresponding topicId (i.e. `oldTopicid` will be null). In this case, 
// when we discover the new topicId, we allow the corresponding leader epoch
// to override the last seen value.
```

Also, would it make sense to move this into the corresponding branch that it applies to?

prat0318 replied:
Ack. Changed the comment as per the suggestion.

> would it make sense to move this into the corresponding branch that it applies to?

Sorry, couldn't get it. Can you elaborate on this please. (Do you mean a separate `if` branch? The current `If` branch deals with separate topicId, so that should be the one we should modify as part of this patch.)

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -394,7 +394,8 @@ else if (metadata.error() == Errors.TOPIC_AUTHORIZATION_FAILED)
-            if (topicId != null && oldTopicId != null && !topicId.equals(oldTopicId)) {
+            // oldTopicId can be null (when metadata is fetched during topic recreation), update the metadata in that case as well.
+            if (topicId != null && !topicId.equals(oldTopicId)) {
                 // If both topic IDs were valid and the topic ID changed, update the metadata
```
nit: this comment is now slightly inaccurate.

prat0318 replied:
updated.

hachikuji commented on core/src/test/scala/integration/kafka/api/ProducerSendWhileDeletionTest.scala:
```diff
@@ -0,0 +1,86 @@
+import java.nio.charset.StandardCharsets
+
+
+class ProducerSendWhileDeletionTest extends BaseProducerSendTest {
```
Out of curiosity, why does this need a new class (as opposed to adding to `PlaintextProducerTest` for example)?

prat0318 replied:
I earlier thought of putting it in `PlaintextProducerTest` but the new case needs 2 as RF for reassignments to take place (to increase leader epoch) and disabling `AutoLeaderRebalance`. Modifying the configs for the whole test class would affect the other tests, so decided to go for a new test class.

dajac replied:
It might be better to extend `IntegrationTestHarness` instead of `BaseProducerSendTest` in this case. `BaseProducerSendTest` contains unit test that will be ran as well as the moment. I don't think that we have to run them with this configuration, or do we?

prat0318 replied:
Agreed that extending `BaseProducerSendTest` would run unnecessary tests. Changed now to extend from `IntegrationTestHarness` and verified that the test fails w/o the fix and passes after the fix.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:
```diff
@@ -381,13 +381,12 @@ public void testEpochUpdateOnChangedTopicIds() {
-        // We should treat an added topic ID as though it is the same topic. Handle only when epoch increases.
-        // Don't update to an older one
-        metadataResponse = RequestTestUtils.metadataUpdateWithIds("dummy", 1, Collections.emptyMap(), Collections.singletonMap("topic-1", 1), _tp -> 1, topicIds);
+        // If the Older topic Id is null, we should go with the new TopicId as the leader Epoch
```
It might be worthwhile having a separate case which goes through the sequence described in the jira. Basically this:
1. Receive metadata response with topicID A.
2. Receive metadata response with UNKNOWN_TOPIC error.
3. Receive metadata response with topicID B.

prat0318 replied:
added a new testcase with the suggested flow. Old test-case still needs to be changed as that case fails now, so modified the case as per the new changes.

dajac replied:
nit: `Older` -> `older`; `topic Id` -> `topic ID`; `TopicId` -> `topic ID`; `Epoch` -> `epoch`.

hachikuji reviewed (commented):
Thanks for the patch! I left a few comments for consideration.

prat0318 committed:
Add review suggestions

prat0318 commented:
@hachikuji Made the suggested changes. Please review.

prat0318 commented:
@hachikuji Gentle bump on the review. PTAL.

prat0318 commented:
@hachikuji @jolshan Bump on the review.

dajac commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -394,8 +394,12 @@ else if (metadata.error() == Errors.TOPIC_AUTHORIZATION_FAILED)
+            // Between the time that a topic is deleted and re-created, the client may lose
+            // track of the corresponding topicId (i.e. `oldTopicId` will be null). In this case,
+            // when we discover the new topicId, we allow the corresponding leader epoch
+            // to override the last seen value.
```
I think that @hachikuji wanted to move this comment block into the `if` branch below, perhaps right after `If the new topic ID is valid and different from the last seen topic ID, update the metadata`.

prat0318 replied:
moved as per the suggestion.

dajac commented on clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:
```diff
@@ -372,6 +372,31 @@ public void testUpdateLastEpoch() {
+    @Test
+    public void testEpochUpdateAfterTopicDeletion() {
+        TopicPartition tp = new TopicPartition("topic-1", 0);
+        Map<String, Uuid> topicIds = Collections.singletonMap("topic-1", Uuid.randomUuid());
```
nit: Would it make sense to move this line right after `Start with a topic A with a topic ID foo` as it goes together with the metadata created there?

prat0318 replied:
ack.

dajac commented on clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:
```diff
@@ -372,6 +372,31 @@ public void testUpdateLastEpoch() {
+        MetadataResponse metadataResponse = emptyMetadataResponse();
+        metadata.updateWithCurrentRequestVersion(metadataResponse, false, 0L);
+
+        // Start with a topic A with a topic ID foo
```
nit: `foo` reads a bit weird here. I would just remove it. Similarly, `topic A` is weird because the topic is name `topic`.

prat0318 replied:
agreed, modified accordingly.

dajac commented on clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:
```diff
@@ -372,6 +372,31 @@ public void testUpdateLastEpoch() {
+        assertEquals(Optional.of(10), metadata.lastSeenLeaderEpoch(tp));
+
+        // Topic A is now deleted so Response contains an Error. LeaderEpoch should still return maintain Old value
+        metadataResponse = RequestTestUtils.metadataUpdateWith("dummy", 1, Collections.singletonMap("topic-1", Errors.UNKNOWN_TOPIC_OR_PARTITION), new HashMap<>());
```
nit: Could we use `Collections.emptyMap()`?

prat0318 replied:
ack

dajac commented on clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:
```diff
@@ -372,6 +372,31 @@ public void testUpdateLastEpoch() {
+        metadata.updateWithCurrentRequestVersion(metadataResponse, false, 1L);
+        assertEquals(Optional.of(10), metadata.lastSeenLeaderEpoch(tp));
+
+        // Topic A is now deleted so Response contains an Error. LeaderEpoch should still return maintain Old value
```
nit: Should `return` be removed?

prat0318 replied:
removed.

dajac commented on core/src/test/scala/integration/kafka/api/ProducerSendWhileDeletionTest.scala:
```diff
@@ -0,0 +1,86 @@
+
+class ProducerSendWhileDeletionTest extends BaseProducerSendTest {
+
+
```
nit: One empty line could be removed.

prat0318 replied:
removed

dajac commented on core/src/test/scala/integration/kafka/api/ProducerSendWhileDeletionTest.scala:
```diff
@@ -0,0 +1,86 @@
+    def testSendWithTopicDeletionMidWay(): Unit = {
+        val numRecords = 10
+
+        // create topic with leader as 0 for the 2 partitions.
```
nit: Could we start all comment with a capital letter to be consistent?

dajac commented on core/src/test/scala/integration/kafka/api/ProducerSendWhileDeletionTest.scala:
```diff
@@ -0,0 +1,86 @@
+        TestUtils.verifyTopicDeletion(zkClient, topic, 2, servers)
+
+        // Producer should be able to send messages even after topic gets deleted and auto-created
+        assertEquals(topic, producer.send(new ProducerRecord(topic, null, ("value").getBytes(StandardCharsets.UTF_8))).get.topic())
```
nit: The parenthesis around `value` could be removed.

prat0318 replied:
ack

dajac commented on core/src/test/scala/integration/kafka/api/ProducerSendWhileDeletionTest.scala:
```diff
@@ -0,0 +1,86 @@
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+
```
nit: Empty line could be removed.

prat0318 replied:
removed.

dajac commented on core/src/test/scala/integration/kafka/api/ProducerSendWhileDeletionTest.scala:
```diff
@@ -0,0 +1,86 @@
+
+        val producer = createProducer(brokerList, maxBlockMs = 5 * 1000L, deliveryTimeoutMs = 20 * 1000)
+
+        (1 to numRecords).map { i =>
```
nit: Using `foreach` instead of `map` would be better here.

prat0318 replied:
done.

dajac commented on core/src/test/scala/integration/kafka/api/ProducerSendWhileDeletionTest.scala:
```diff
@@ -0,0 +1,86 @@
+        assertEquals(topic, producer.send(new ProducerRecord(topic, null, ("value").getBytes(StandardCharsets.UTF_8))).get.topic())
+    }
+
+
```
nit: One empty line could be removed here as well.

prat0318 replied:
removed.

dajac reviewed (commented):
@prat0318 Thanks for the PR! I do agree with @hachikuji that the workaround looks reasonable. I have left some comments below (mainly nits).

prat0318 committed:
Add Review suggestions

prat0318 committed:
Capitalize start of the comment lines

prat0318 committed:
Fix build for Scala 12

dajac commented on clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:
```diff
@@ -372,6 +372,30 @@ public void testUpdateLastEpoch() {
+        MetadataResponse metadataResponse = emptyMetadataResponse();
+        metadata.updateWithCurrentRequestVersion(metadataResponse, false, 0L);
+
+        Map<String, Uuid> topicIds = Collections.singletonMap("topic-1", Uuid.randomUuid());
```
nit: We could have kept the comment before the block to be consistent with the other two blocks. I just wanted to rework it a bit.

dajac commented on clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:
```diff
@@ -372,6 +372,30 @@ public void testUpdateLastEpoch() {
+        metadata.updateWithCurrentRequestVersion(metadataResponse, false, 1L);
+        assertEquals(Optional.of(10), metadata.lastSeenLeaderEpoch(tp));
+
+        // Create topic-1 again but this time with a topic ID bar. LeaderEpoch should be updated to new even if lower.
```
nit: `Create topic-1 again but this time with a topic ID.`?

dajac commented on core/src/test/scala/integration/kafka/api/ProducerSendWhileDeletionTest.scala:
```diff
@@ -0,0 +1,83 @@
+            new TopicPartition(topic, 1) -> Seq(1, 0)
+        )
+
+        // Change leader to 1 for both the partitions to increase leader Epoch from 0 -> 1
```
nit: `Epoch` -> `epoch`.

dajac commented:
@prat0318 Thanks for the update. I left a few more nits to fix typos. Would you have time to quickly address them?

prat0318 committed:
Add review suggestions

prat0318 commented:
> @prat0318 Thanks for the update. I left a few more nits to fix typos. Would you have time to quickly address them?

@dajac Thanks again for the review. I have addressed the suggestions. Please re-review.

dajac reviewed (approved):
LGTM, thanks for the patch!

dajac commented:
Merged to trunk, 3.1, 3.0 and 2.8.
